### ISSUE TITLE
[Umbrella] Analyzer, Planner migration

### ISSUE THREAD

**Issue by kitaisreal** [ClickHouse/ClickHouse#42648]
After https://github.com/ClickHouse/ClickHouse/pull/31796 was merged now we have new infrastructure for query analysis and planning. There are a lot of things that new infrastructure brings to us:
1. Better performance of queries that spend considerable time during query analysis, planning stage. Better performance of queries with a lot of JOINs, a lot of subqueries. Example: https://github.com/ClickHouse/ClickHouse/issues/39996.
2. A lot of new SQL features that are now possible to implement (better scalar subqueries, correlated subqueries).
3. Better support for JOINs (Support for indexes, FINAL, SAMPLE).
4. Ability to implement full featured query optimizer on top of new infrastructure.
5. Better usability, better exception messages.
6. Full SQL specificiation of our SQL extensions is now possible to write.

This ticket will contain tasks that are necessary to implement before we can enable new infrastructure by default.

## Analyzer
- [x] SELECT (compound_expression).*, (compound_expression).COLUMNS are not supported on parser level. Developer @evillique. #42884
- [x] SELECT a.b.c.*, a.b.c.COLUMNS. Qualified matcher where identifier size is greater than 2 are not supported on parser level. Developer @evillique. #42884
- [ ] Support function identifier resolve from parent query scope, if lambda in parent scope does not capture any columns.
- [x] Support group_by_use_nulls. Developer @novikd. #45910
- [x] Support cache for scalar subqueries. Developer @SmitaRKulkarni. #43640
- [x] Support query tree in table functions. Developer @kitaisreal. #46471
- [x] Support `SELECT * FROM table AS table` for table expressions. Developer @kitaisreal. #43089.
- [x] Support typo corrections for identifiers. Support typo corrections for functions. Developer @kitaisreal. #43002.
- [x] Support storage Merge. Developer @kitaisreal. #46471
- [x] Support storage Materialized View. Developer @kitaisreal. #46471
- [x] Support storage Live View. Developer @kitaisreal. #46471
- [ ] Support storage Window View.
- [x] Support transform query for external database. Developer @vdimir. #47316
- [x] Join columns refactoring. Developer @novikd #58317 @vdimir #59007
- [ ] Constants refactoring. Developer @kitaisreal #59523
- [x] Execution names refactoring. Developer @novikd #53705

## QueryTreePassManager
- [x] Support logical expressions optimizer. Developer @antonio2368 https://github.com/ClickHouse/ClickHouse/pull/46644
- [x] Support fuse sum count optimize_fuse_sum_count_avg, optimize_syntax_fuse_functions. Developer @vdimir. #42689
- [x] Support setting convert_query_to_cnf. Developer @antonio2368 https://github.com/ClickHouse/ClickHouse/pull/47617
- [x] Support setting optimize_using_constraints. Developer @antonio2368 https://github.com/ClickHouse/ClickHouse/pull/47617
- [x] Support setting optimize_substitute_columns. Developer @antonio2368 https://github.com/ClickHouse/ClickHouse/pull/47617
- [x] Support GROUP BY injective function elimination. Developer @novikd #58919
- [x] Support GROUP BY functions of other keys elimination. Developer @novikd. #43261
- [x] Support setting optimize_move_functions_out_of_any. Developer @JackyWoo. #52082
- [x] Support setting optimize_aggregators_of_group_by_keys. Developer @JackyWoo @novikd. #52230
- [x] Support setting optimize_redundant_functions_in_order_by. Developer @novikd. #42970
- [ ] Support setting optimize_monotonous_functions_in_order_by. Developer @novikd.
- [x] Support setting optimize_if_transform_strings_to_enum. Developer @antonio2368. https://github.com/ClickHouse/ClickHouse/pull/43999
- [x] Support settings.optimize_or_like_chain. Developer @novikd. #42797
- [ ] Add optimizations based on function semantics. Example: SELECT * FROM test_table WHERE id != id. (id is not nullable column).
- [x] Support CROSS JOIN to INNER JOIN rewrite. Developer @vdimir. #46408
- [x] Support _shard_num identifier to shardNum() function rewriting. Developer @kitaisreal. #47101
- [x] Support remove unused projection columns in subqueries. Developer @kitaisreal. #56499

# Planner
- [x] Support JOIN with JOIN engine. Developer @vdimir. #43897
- [x] JOIN drop unnecessary columns after ON, USING section. @vdimir #44545
- [x] Support RBAC. Developer @kitaisreal. #47622
- [x] Support distributed query processing. Developer @kitaisreal. #44540 #47383
- [x] Support projections. Developer @KochetovNicolai.
- [x] Support UNION storage limits. Support storage limits in general. Need to check all places in InterpreterSelectQuery. Developer @yakov-olkhovskiy. #46306
- [x] Support indexes. Support indexes for JOINS. Developer @kitaisreal. #43341 #44097
- [ ] Support experimental hypothesis indexes. Developer @antonio2368. https://github.com/ClickHouse/ClickHouse/pull/48381
- [ ] Support experimental ANN indexes. Developer @kitaisreal.
- [x] Support indexes for IN function. Developer @kitaisreal.
- [x] Support settings limit and offset. Developer @yakov-olkhovskiy. #46632
- [x] Support trivial count optimization. Developer @devcrafter. #46923
- [x] StorageMerge refactoring. Developer @novikd 

# QueryPlan optimizations

- [x] Support read in order optimizations. Developer @KochetovNicolai. #42829 #43592
- [x] Support ORDER BY read in order optimization. Developer @KochetovNicolai. #42829
- [x] Support GROUP BY read in order optimization. Developer @KochetovNicolai. #43592
- [x] Support PREWHERE. Developer @kitaisreal. #47622
- [x] Support setting optimize_duplicate_order_by_and_distinct. Developer @devcrafter. ~#43229~ #43905 #44176

# Backward compatibility
- [x] Added setting `single_join_prefer_left_table`. Developer @kitaisreal. #46471

# Additional tasks
- [ ] #58027
- [ ] Tune performance of Analyzer using different low level optimizations.
- [ ] Add performance tests in our CI/CD for big queries with a lot of JOINS.
- [ ] Add performance tests for high concurrent queries workload scenario.

# Broken tests

#### Grouped
- [x] distributed_unused_shards @kitaisreal @KochetovNicolai `00754_distributed_optimize_skip_select_on_unused_shards` `00754_distributed_optimize_skip_select_on_unused_shards_with_prewhere` `01071_force_optimize_skip_unused_shards` `01072_optimize_skip_unused_shards_const_expr_eval` `01211_optimize_skip_unused_shards_type_mismatch` `01213_optimize_skip_unused_shards_DISTINCT` `01247_optimize_distributed_group_by_sharding_key_dist_on_dist` `01270_optimize_skip_unused_shards_low_cardinality` `01319_optimize_skip_unused_shards_nesting` `01527_dist_sharding_key_dictGet_reload` `01528_allow_nondeterministic_optimize_skip_unused_shards` `01563_distributed_query_finish` `01756_optimize_skip_unused_shards_rewrite_in` `01757_optimize_skip_unused_shards_limit` `01758_optimize_skip_unused_shards_once` `01759_optimize_skip_unused_shards_zero_shards` `01901_in_literal_shard_prune` `01930_optimize_skip_unused_shards_rewrite_in`  https://github.com/ClickHouse/ClickHouse/pull/53282
- [x] projections `01576_alias_column_rewrite` `01710_projection_additional_filters` #54128
- [x] subquery_in_index @KochetovNicolai `01700_system_zookeeper_path_in 00562_in_subquery_merge_tree` `00673_subquery_prepared_set_performance` `00945_bloom_filter_index` `00981_in_subquery_with_tuple` `01583_const_column_in_set_index` `01585_use_index_for_global_in` `01585_use_index_for_global_in_with_null` `01681_bloom_filter_nullable_column` `01739_index_hint`
- [x] rewrite `00597_push_down_predicate_long`(https://github.com/ClickHouse/ClickHouse/pull/49551)  `01655_plan_optimizations` (already fixed), `01655_plan_optimizations_optimize_read_in_window_order`(https://github.com/ClickHouse/ClickHouse/pull/49565) `01951_distributed_push_down_limit`, `02343_aggregation_pipeline` (already fixed), `02377_optimize_sorting_by_input_stream_properties_explain`(https://github.com/ClickHouse/ClickHouse/pull/49943),  `02493_inconsistent_hex_and_binary_number` `02511_complex_literals_as_aggregate_function_parameters` (https://github.com/ClickHouse/ClickHouse/pull/49800) @devcrafter 
- [ ] storage_materialized_view `01155_rename_move_materialized_view` `01268_mv_scalars` `01947_mv_subquery` `02139_MV_with_scalar_subquery` `02174_cte_scalar_cache_mv` `02479_race_condition_between_insert_and_droppin_mv` @yakov-olkhovskiy 
- [x] storage_merge `01083_expressions_in_engine_arguments` `01214_test_storage_merge_aliases_with_where` `01560_merge_distributed_join` `01925_test_storage_merge_aliases` `02402_merge_engine_with_view` `02575_merge_prewhere_different_default_kind` `02763_row_policy_storage_merge_alias` @novikd https://github.com/ClickHouse/ClickHouse/pull/50894
- [x] join_use_nulls `01142_join_lc_and_nullable_in_key` `01142_merge_join_lc_and_nullable_in_key` `01476_right_full_join_switch` `01477_lc_in_merge_join_left_key` `02024_join_on_or_long` https://github.com/ClickHouse/ClickHouse/pull/48893 and https://github.com/ClickHouse/ClickHouse/pull/49359 @vdimir
- [x] join_types `01049_join_low_card_bug_long` `01353_low_cardinality_join_types` `01721_join_implicit_cast_long` `02242_join_rocksdb` `02267_join_dup_columns_issue36199` https://github.com/ClickHouse/ClickHouse/pull/49555 @vdimir
- [ ] `01062_pm_all_join_with_block_continuation` @vdimir https://github.com/ClickHouse/ClickHouse/pull/56458
- [x] `02481_merge_array_join_sample_by` https://github.com/ClickHouse/ClickHouse/pull/49784 @vdimir 
- [x] distributed_different_databases `01152_cross_replication` `01232_extremes` `01487_distributed_in_not_default_db` `01615_random_one_shard_insertion` @novikd #50097 #50587
- [x] invalid_set `00700_decimal_compare` `02421_decimal_in_precision_issue_41125` `02428_decimal_in_floating_point_literal` @vdimir #48754
- [x] transform_query_for_external_database `01086_odbc_roundtrip` #54133
- [x] partition_pruning `01540_verbatim_partition_pruning` `01748_partition_id_pruning` https://github.com/ClickHouse/ClickHouse/pull/54185
- [x] subquery_remove_unused_columns `00593_union_all_assert_columns_removed` `01455_shard_leaf_max_rows_bytes_to_read` `01586_columns_pruning`  @kitaisreal #56499
- [x] key_condition `01651_bugs_from_15889` `01786_explain_merge_tree` `01889_key_condition_function_chains` https://github.com/ClickHouse/ClickHouse/pull/56211
- [x] additional_filters `02346_additional_filters_distr`
- [x] set_index `00838_unique_index` `00979_set_index_not` https://github.com/ClickHouse/ClickHouse/pull/52754 @devcrafter 
- [x] use_structure_from_insertion_table_in_table_functions `02458_use_structure_from_insertion_table`
- [x] storage_key_value `01504_rocksdb` `02375_rocksdb_with_filters` https://github.com/ClickHouse/ClickHouse/pull/56391
- [x] transactions `01173_transaction_control_queries` `02345_implicit_transaction` @kitaisreal
- [ ] optimize_using_constraints `01624_soft_constraints` https://github.com/ClickHouse/ClickHouse/pull/48381
- [x] read_in_order `00940_order_by_read_in_order_query_plan` https://github.com/ClickHouse/ClickHouse/pull/56532 @kitaisreal 
- [ ] dependend_objects `01747_join_view_filter_dictionary` @kitaisreal
- [x] projection_names `02426_orc_bug` #56514 @kitaisreal 
- [x] alias_columns `02494_optimize_group_by_function_keys_and_alias_columns`
- [x] parameterized_view `02428_parameterized_view` `02818_parameterized_view_with_cte_multiple_usage` @novikd @SmitaRKulkarni #54211
- [x] allow_compound_columns `00927_asof_joins` @vdimir
- [x] allow_aggregate_partitions_independently `02521_aggregation_by_partitions` https://github.com/ClickHouse/ClickHouse/pull/56254
- [x] join_with_constants `02382_join_and_filtering_set` @vdimir
- [x] cte_distributed `02341_global_join_cte`
- [x] remove_join `01890_materialized_distributed_join` @vdimir https://github.com/ClickHouse/ClickHouse/pull/55741
- [ ] distributed_conversion `01268_shard_avgweighted` `/// Not found column avgWeighted(100_Int8, -1_Int8) in block. There are only columns: avgWeighted(100_UInt8, -1_Int8)` @kitaisreal
- [x] invalid_view_subquery_parsing `02366_union_decimal_conversion` https://github.com/ClickHouse/ClickHouse/pull/56274
- [ ] prewhere `00717_merge_and_distributed` Virtual columns for merge : `Column _table not found in table _table_function.merge`
- [x] any_join_const `02000_join_on_const` @vdimir
- [x] recursive_cte `01495_subqueries_in_with_statement` Infinit cte substitution? `WITH test1 AS (SELECT i + 1, j + 1 FROM test1) SELECT * FROM test1` -> `Too deep subqueries`
- [ ] used_functions_log `01656_test_query_log_factories_info` We create some helper functions (`or`, `if`, `multiif`, `multiMatchAny`), in passes even we don't need it. Alos there is a difference in `CRC32` case.
- [x] memory_bound_merging `02404_memory_bound_merging` https://github.com/ClickHouse/ClickHouse/pull/56419
- [x] row_policy `02131_used_row_policies_in_query_log` #56531 @kitaisreal 
- [ ] distributed `00223_shard_distributed_aggregation_memory_efficient` `select randConstant(), randConstant(), randConstant()` -> `Block structure mismatch`
- [x] asof_join_nullable `01428_nullable_asof_join`  https://github.com/ClickHouse/ClickHouse/pull/49826 @vdimir
- [x] storage_s3 `02302_s3_file_pruning` @vdimir https://github.com/ClickHouse/ClickHouse/pull/56668
- [x] distinct_in_order `02317_distinct_in_order_optimization_explain` (https://github.com/ClickHouse/ClickHouse/pull/48669) @devcrafter 
- [x] distributed_aggregation_memory_efficient `01231_distributed_aggregation_memory_efficient_mix_levels`
- [ ] cast_null `01761_cast_to_enum_nullable` @kitaisreal
- [x] grouping_sets `02554_fix_grouping_sets_predicate_push_down` https://github.com/ClickHouse/ClickHouse/pull/56595
- [x] input_function `00952_input_function`
- [x] sorting in distributed queries `02001_shard_num_shard_count` `test_parallel_replicas_skip_shards/test.py::test_skip_unavailable_shards`
- [x] invalid_aggregation `02352_grouby_shadows_arg`
- [x] merge_tree_index `01064_incremental_streaming_from_2_src_with_feedback`
- [ ] storage_buffer `01584_distributed_buffer_cannot_find_column`
- [ ] big_constants `00725_memory_tracking` @kitaisreal
- [ ] join_column_cast `01925_join_materialized_columns` @vdimir
- [x] `01739_index_hint` `02880_indexHint__partition_id` @KochetovNicolai 

#### Integrations tests
- [x] `test_wrong_db_or_table_name/test.py::test_wrong_table_name`, `test_sql_user_defined_functions_on_cluster/test.py::test_sql_user_defined_functions_on_cluster`, `test_user_defined_object_persistence/test.py::test_persistence`, `test_dictionaries_update_and_reload/test.py::test_reload_after_fail_in_cache_dictionary`, `test_access_for_functions/test.py::test_access_rights_for_function`, `test_select_access_rights`
- [x] `test_profile_events_s3/test.py::test_profile_events`
- [x] `test_mysql_database_engine/test.py::test_mysql_ddl_for_mysql_database`
- [x] `test_executable_table_function/test.py::test_executable_function_input_python`
- [x] `test_cluster_copier`
- [x] `test_backward_compatibility`, `test_distributed_backward_compatability/test.py::test_distributed_in_tuple` distributed between nodes with old and new analyzer
- [x] `test_user_defined_object_persistence/test.py::test_persistence` 
- [x] `test_storage_rabbitmq/test.py::test_rabbitmq_materialized_view` `SELECT * FROM test.rabbitmq GROUP BY (key, value)`
- [x] `test_mutations_with_merge_tree/test.py::test_mutations_with_merge_background_task`
- [x] `test_build_sets_from_multiple_threads/test.py::test_set`

#### Performance tests

Separate issue was created for this topic: #62245.

- [ ] `direct_dictionary`, `set_index` @kitaisreal 
- [ ] `push_down_limit`
- [ ] `injective_functions_inside_uniq`
- [ ] `agg_functions_min_max_any`
- [ ] `norm_distance_float`, `norm_distance`
- [ ] `array_fill` big constants @kitaisreal
- [ ] `logical_functions_medium `
- [ ] `if_transform_strings_to_enum`
- [ ] `distinct_combinator`
- [ ] `arithmetic_operations_in_aggr_func`
- [ ] `aggregate_functions_of_group_by_keys`
- [ ] `order_by_read_in_order`
- [ ] `aggregate_with_serialized_method`
- [ ] `materialized_view_parallelize_output_from_storages`
- [ ] `if_transform_strings_to_enum`
- [ ] `low_cardinality_query`
- [ ] `uniqExactIf`
- [ ] `parallel_index`
- [ ] `rand` rewrite
- [ ] `group_by_fixed_keys` rewrite
- [ ] `logical_functions_small`, `logical_functions_medium`, `logical_functions_large` rewrite
- [ ] `hash_table_sizes_stats` rewrite
- [ ] `joins_in_memory` rewrite

#### Other
- [x] `02713_array_low_cardinality_string`
- [x] `02707_skip_index_with_in`
- [x] `02707_complex_query_fails_analyzer` #49403
- [x] `02680_mysql_ast_logical_err` @vdimir #49362
- [x] `02324_map_combinator_bug`
- [x] `02241_join_rocksdb_bs` @vdimir
- [x] `02003_WithMergeableStateAfterAggregationAndLimit_LIMIT_BY_LIMIT_OFFSET`
- [x] `01626_cnf_fuzz_long` https://github.com/ClickHouse/ClickHouse/pull/49402
- [x] `01115_join_with_dictionary` @vdimir
- [x] `01009_global_array_join_names` @vdimir https://github.com/ClickHouse/ClickHouse/pull/56587
- [x] `00917_multiple_joins_denny_crane` @vdimir
- [x] `00725_join_on_bug_1` @vdimir
- [x] `00636_partition_key_parts_pruning`
- [x] `00261_storage_aliases_and_array_join` https://github.com/ClickHouse/ClickHouse/pull/55618
- [x] `02701_non_parametric_function` #49419
- [x] `01825_type_json_multiple_files`
- [x] `01281_group_by_limit_memory_tracking`
- [x] `02723_zookeeper_name` @vdimir
- [x] `01019_alter_materialized_view_consistent` #49261 https://github.com/ClickHouse/ClickHouse/pull/56246
- [ ]  `01244_optimize_distributed_group_by_sharding_key` `01952_optimize_distributed_group_by_sharding_key`
- [ ] `01287_max_execution_speed` https://github.com/ClickHouse/ClickHouse/issues/56052
- [ ] `02725_agg_projection_resprect_PK` @KochetovNicolai
- [x] `02784_parallel_replicas_automatic_decision_join` @Algunenano (broken test added in https://github.com/ClickHouse/ClickHouse/pull/51692) @KochetovNicolai https://github.com/ClickHouse/ClickHouse/pull/58838

#### Later
- [ ] experimental_ann_index `02354_annoy`

**Comment by kitaisreal** [comment]
Some examples of features that new infrastructure brings us:
Standalone lambdas:
```
WITH x -> * AS lambda SELECT lambda(1)
```
Standalone lambdas carrying.

Transformers are native part of language:
```
SELECT * APPLY x -> (* EXCEPT value) FROM test_table
SELECT arrayMap(x -> test_table.* EXCEPT id APPLY toString, [1,2,3]) FROM test_table;
```
Matchers to compound alias query expressions.
```
SELECT cast(tuple(1, 'Value'), 'Tuple (id UInt64, value String)') AS a, a.id, a.value;
SELECT cast(tuple(1, 'Value'), 'Tuple (id UInt64, value String)') AS a, a.*;
```
Multiple ARRAY JOINs. Example:
```
SELECT t1.id, value1, value2 FROM test_table_join_1 AS t1 ARRAY JOIN [[1,2,3]] AS value1 ARRAY JOIN value1 AS value2;
```
Multiple USING in JOINs statements.
Multiple JOINS allow access to previous table identifiers. Example:
```
SELECT t1.value, t2.value, t3.value FROM test_table_join_1 AS t1
INNER JOIN test_table_join_2 AS t2 ON t1.id = t2.id
INNER JOIN test_table_join_3 AS t3 ON t1.id = t3.id OR t1.id = t2.id;
```
Multiple JOINS export aliases. Example:
```
SELECT t1_id, t1.value, t2_id, t2.value, t3_id, t3.value FROM test_table_join_1 AS t1
INNER JOIN test_table_join_2 AS t2 ON (t1.id AS t1_id) = (t2.id AS t2_id)
INNER JOIN test_table_join_3 AS t3 ON t2_id = (t3.id AS t3_id);
```
JOINS support SAMPLE BY. Example:
```
SELECT t1.id AS t1_id, t2.id AS t2_id, t1._sample_factor AS t1_sample_factor, t2._sample_factor AS t2_sample_factor
FROM test_table_join_1 AS t1 SAMPLE 1/2 INNER JOIN test_table_join_2 AS t2 SAMPLE 1/2 ON t1.id = t2.id;
```
JOINS support FINAL. Example:
```
SELECT t1.id AS t1_id, t2.id AS t2_id, t1.value AS t1_value, t2.value AS t2_value
FROM test_table_join_1 AS t1 FINAL INNER JOIN test_table_join_2 AS t2 FINAL ON t1.id = t2.id;
```

**Comment by jrdi** [comment]
@kitaisreal amazing job here!

I didn't see indexes support in the list of tasks, and just wanted to understand if will be part of this effort.
I see a couple of references to indexes though but looks like are specific for IN clause, and JOINs

**Comment by kitaisreal** [comment]
@jrdi updated tasks. This is actually first prioirity right now, pull request will be created soon.
```
Support indexes. Support indexes for JOINS.
```

**Comment by vdimir** [comment]
May you please check if the analyzer should support this query?

```sql
SELECT * FROM (SELECT sum(a) FROM (SELECT 1 as a)) SETTINGS allow_experimental_analyzer = 1;

DB::Exception: Aggregate function sum(a) is found in JOIN TREE in query. (ILLEGAL_AGGREGATION)
```

**Comment by kitaisreal** [comment]
@vdimir already fixed https://github.com/ClickHouse/ClickHouse/pull/42865.

**Comment by qoega** [comment]
I will collect here failures with analyser:
- #43029

**Comment by qoega** [comment]
When we will enable analyzer by default it will close #12264

**Comment by novikd** [comment]
> Support setting optimize_monotonous_functions_in_order_by.

It seems that this optimization doesn't worth reimplementation. Right now it's disabled by default because it can break read in order optimization. It's impossible to implement it in a good way on top of AST or Query Tree because this kind of optimization requires the knowledge of input datastream order.

I discussed it with @KochetovNicolai and he is okay with removing the optimization. I'll think more if we can reimplement it in some other place, but most likely I'll remove it.

**Comment by qoega** [comment]
arrayMap is not considered constant in cte. Is it known?
https://pastila.nl/?04eaaa06/065ecaa5a226f4f7a738c65714cb75d3

**Comment by UnamedRus** [comment]
It's kinda known problem:

https://github.com/ClickHouse/ClickHouse/issues/14980

**Comment by jrdi** [comment]
I guess this is covered by

>  Support distributed query processing. Developer @kitaisreal.

But just in case, queries using cluster operations 

```sql
SELECT *
FROM cluster('default', system.query_log)
LIMIT 1
SETTINGS allow_experimental_analyzer = 1
```

Are throwing the following error when using the analyzer:

```
Received exception from server (version 22.12.1):
Code: 36. DB::Exception: Received from clickhouse-01:9000. DB::Exception: Element of set in IN, VALUES or LIMIT or aggregate function parameter is not a constant expression (result column not found): system.query_log. (BAD_ARGUMENTS)
```

**Comment by qoega** [comment]
https://github.com/ClickHouse/ClickHouse/issues/43892

**Comment by qoega** [comment]
https://github.com/ClickHouse/ClickHouse/issues/44490

**Comment by zhanglistar** [comment]
@kitaisreal Could you explain a little more things about how to do distributed query processing, do you will split SQL into multiple fragment to execution on multiple nodes or the style now clickhouse have? Thanks a lot.

**Comment by kitaisreal** [comment]
@zhanglistar for now during migration process it is implemented with sending SQL to shard nodes. Check out https://github.com/ClickHouse/ClickHouse/pull/44540.
After migration to new query analysis and planning architecture will be finished. We want to serialize query plan on initiator and send it to shard nodes, so it will be easier to maintain and extend.

**Comment by vdimir** [comment]
- [ ] `01049_join_low_card_bug_long`

`Identifier t.t nested path t cannot be resolved from type UInt8 from subquery with name a. In scope SELECT t.t FROM (SELECT 1 AS t) AS a FULL OUTER JOIN (SELECT 1 AS t) AS t USING (t). (UNKNOWN_IDENTIFIER)`

```sql
select t.t from (SELECT 1 as t) as a
FULL JOIN (SELECT 1 as t) as t USING (t);
```

- [ ] `02242_join_rocksdb`

`Block structure mismatch in (columns with identical name must have identical structure) stream: different types: key UInt32 UInt32(size = 0) key String String(size = 0). (AMBIGUOUS_COLUMN_NAME)`

```sql

CREATE TABLE rdb (key UInt32, value String) ENGINE = EmbeddedRocksDB PRIMARY KEY (key);
INSERT INTO rdb VALUES (1, 'val1');

SELECT * FROM (SELECT 1 as key) as t2
INNER JOIN rdb ON rdb.key == t2.key
SETTINGS join_algorithm = 'direct';
```

**Comment by helifu** [comment]
Look forward to new progress ^_^

**Comment by JackyWoo** [comment]
@kitaisreal I summited some PRs for `QueryTreePassManager ` tasks, details in https://github.com/ClickHouse/ClickHouse/issues/52081.

**Comment by Algunenano** [comment]
`optimize_rewrite_sum_if_to_count_if` can be enabled with the analyzer:

```patch
diff --git a/tests/queries/0_stateless/01646_rewrite_sum_if_bug.sql b/tests/queries/0_stateless/01646_rewrite_sum_if_bug.sql
index 3e6a7b92dbf..ba9e51fc0f9 100644
--- a/tests/queries/0_stateless/01646_rewrite_sum_if_bug.sql
+++ b/tests/queries/0_stateless/01646_rewrite_sum_if_bug.sql
@@ -1,3 +1,5 @@
+SET optimize_rewrite_sum_if_to_count_if = 1;
+SET allow_experimental_analyzer = 1;
 DROP TABLE IF EXISTS t;
 create table t( s String ) Engine=Memory as select arrayJoin (['a','b','c']);
 
@@ -12,11 +14,11 @@ DROP TABLE IF EXISTS test_alias;
 CREATE TABLE test_alias(`a` Int64, `b` Int64, `c` Int64, `day` Date, `rtime` DateTime) ENGINE = Memory
 as select 0, 0, 0, '2022-01-01', 0 from zeros(10);
 
-WITH 
-    sum(if((a >= 0) AND (b != 100) AND (c = 0), 1, 0)) AS r1, 
-    sum(if((a >= 0) AND (b != 100) AND (c > 220), 1, 0)) AS r2 
-SELECT 
-    (intDiv(toUInt32(rtime), 20) * 20) * 1000 AS t, 
+WITH
+    sum(if((a >= 0) AND (b != 100) AND (c = 0), 1, 0)) AS r1,
+    sum(if((a >= 0) AND (b != 100) AND (c > 220), 1, 0)) AS r2
+SELECT
+    (intDiv(toUInt32(rtime), 20) * 20) * 1000 AS t,
     (r1 * 100) / (r1 + r2) AS m
 FROM cluster('test_cluster_two_shards', currentDatabase(), test_alias)
 WHERE day = '2022-01-01'
```

**Comment by alexey-milovidov** [comment]
We switched to the analyzer three months ago already.
